### PR TITLE
States refactor, removed factory from states.

### PR DESCRIPTION
**Pull request by KharchenkoMaks**
- Removed loop dependency between states factory and states.
- Removed states factory from states, states now return StateType enum and StateMachine is creating next steps using factory.
- Made StatesFactory general `interface`, moved implementation to `LazyStatesFactory`
- Made `context` of `StateMachine` to be `unique_ptr`
- Renamed `SetLabelState` to `AddLabelState`

### CODE REVIEW

File: src/view/states/task_input/SetLabelState.cpp [hunk 1]
```diff
@@ -5,22 +5,28 @@
...
     if (!task_id.has_value()){
-        factory_.lock()->GetConsolePrinter()->WriteError("Incorrect task id was given, try again!");
-        return factory_.lock()->GetNextState(*this, StatesFactory::MoveType::ERROR);
+        printer_->WriteError("Incorrect task id was given, try again!");
+        return error_state_;
     }
 
-    std::string label_to_set = console_io::util::GetUserInput("Label", *factory_.lock()->GetConsolePrinter(), *factory_.lock()->GetConsoleReader());
+    std::string label_to_set = console_io::util::GetUserInput("Label", *printer_, *reader_);
     context.SetTaskId(task_id.value());
```
Comment (AntonApraksin) [inline src/view/states/task_input/SetLabelState.cpp]: You have already checked if `task_id` contains a value in `22`. Just dereference here.

**Review by AntonApraksin — APPROVED** [review]

File: src/view/states/State.h [hunk 2]
```diff
@@ -6,19 +6,40 @@
...
 class State {
 public:
-    virtual std::shared_ptr<State> Execute(StateContext&) = 0;
+    virtual StateType Execute(StateContext&) = 0;
 
 public:
     virtual ~State() = default;
 };
 
+enum class StateType {
```
Comment (antonovchtd) [inline src/view/states/State.h:23]: StateType feels more like implementation, not interface. Looks strange to bring concrete steps names to Step.h, but I don't have any suggestions at the moment

File: src/view/states/task_input/AddTaskState.h [hunk 3]
```diff
@@ -6,19 +6,24 @@
...
 #include "user_interface/StateContext.h"
+#include "view/user_interface/ConsoleStateMachine.h"
 
 #include <memory>
 #include <optional>
 
 class AddTaskState : public State {
 public:
-    explicit AddTaskState(const std::shared_ptr<StatesFactory>& factory);
+    AddTaskState(StateType next_state,
```
Comment (antonovchtd) [inline src/view/states/task_input/AddTaskState.h:18]: AddTaskState, as any, could carry error_state in general - what if the task was not added? (task id collision, network issues, permissions, etc)

**Review by antonovchtd — APPROVED** [review]

File: src/view/states/task_input/AddTaskState.h (hunk 3 above)
Comment (KharchenkoMaks) [reply src/view/states/task_input/AddTaskState.h:18]: Each state's responsibility is ask user for information and create a `command`. Processing of errors you've mentioned isn't responsibility of states.

**Review by MichaelKupriianov — APPROVED** [review]

**Review by vladimir-moiseiev — APPROVED** [review]

**Pull request merged**